authentication/authentication.go

```diff
@@ -0,0 +1,150 @@
+	if err != nil {
+		return "", &TokenError{Desc: "Error creating JWT signer", From: err}
+	}
+
+	token, err := jwt.SignedAndEncrypted(signer, encrypter).Claims(rawClaims).CompactSerialize()
+	if err != nil {
+		return "", &TokenError{Desc: "Error signing and encrypting JWT", From: err}
+	}
+
+	return token, nil
+}
+
+// PublicKeyResult is a wrapper for the public key and the kid that identifies it
+type PublicKeyResult struct {
+	key *rsa.PublicKey
+	kid string
+}
+
+// PrivateKeyResult is a wrapper for the private key and the kid that identifies it
+type PrivateKeyResult struct {
+	key *rsa.PrivateKey
+	kid string
+}
+
+func loadSigningKey(signingKeyPath string) (*PrivateKeyResult, *KeyLoadError) {
+
+	keyData, err := ioutil.ReadFile(signingKeyPath)
+	if err != nil {
+		return nil, &KeyLoadError{Op: "read", Err: "Failed to read signing key from file: " + signingKeyPath}
+	}
+
+	block, _ := pem.Decode(keyData)
+	privateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, &KeyLoadError{Op: "parse", Err: "Failed to parse signing key from PEM"}
+	}
+
+	PublicKey, err := x509.MarshalPKIXPublicKey(&privateKey.PublicKey)
+	if err != nil {
+		return nil, &KeyLoadError{Op: "marshal", Err: "Failed to marshal public key"}
+	}
+
+	pubBytes := pem.EncodeToMemory(&pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: PublicKey,
+	})
+
+	kid := fmt.Sprintf("%x", sha1.Sum(pubBytes))
+
+	return &PrivateKeyResult{privateKey, kid}, nil
+}
+
+func loadEncryptionKey(encryptionKeyPath string) (*PublicKeyResult, *KeyLoadError) {
+
+	keyData, err := ioutil.ReadFile(encryptionKeyPath)
+	if err != nil {
+		return nil, &KeyLoadError{Op: "read", Err: "Failed to read encryption key from file: " + encryptionKeyPath}
+	}
+
+	block, _ := pem.Decode(keyData)
+	pub, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, &KeyLoadError{Op: "parse", Err: "Failed to parse encryption key PEM"}
+	}
+
+	kid := fmt.Sprintf("%x", sha1.Sum(keyData))
+
+	publicKey, ok := pub.(*rsa.PublicKey)
+	if !ok {
+		return nil, &KeyLoadError{Op: "cast", Err: "Failed to cast key to rsa.PublicKey"}
+	}
+
+	return &PublicKeyResult{publicKey, kid}, nil
+}
+
+// TokenError describes an error that can occur during JWT generation
+type TokenError struct {
+	// Err is a description of the error that occurred.
+	Desc string
+
+	// From is optionally the original error from which this one was caused.
+	From error
+}
+
+// KeyLoadError describes an error that can occur during key loading
+type KeyLoadError struct {
+	// Op is the operation which caused the error, such as
+	// "read", "parse" or "cast".
+	Op string
+
+	// Err is a description of the error that occurred during the operation.
+	Err string
+}
+
+func (e *KeyLoadError) Error() string {
+	if e == nil {
+		return "<nil>"
+	}
+	return e.Op + ": " + e.Err
+}
+
+func (e *TokenError) Error() string {
+	if e == nil {
+		return "<nil>"
+	}
+	err := e.Desc
```

Comment:
`err` should be reserved for holding the actual error (i.e. replace `e` in this function). If it's not an error type then it should have a different name.

Reply:
Done